### PR TITLE
Replace login-related makos with vue components

### PR DESCRIPTION
**Pull request by guerler**
This PR replaces the login mako with a vue component and connects it with api-like login endpoint in the user controller for Galaxy. It also replaces the change password, and forgotten password options, and the open id login form. Legacy code and makos have been moved to the toolshed. It simplifies the upcoming replacement of the registration mako. 
 
Todos:
 - [x] Redirect for change password view
 - [x] Add test cases
 - [x] Fix/check activation link

### CODE REVIEW

**Comment by bgruening** [comment]
Nice!

File: lib/galaxy/managers/users.py
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]: I think you can now remove the comment line 31.

**Comment by dannon** [comment]
I'm still concerned about the duplication of code here, xref #6512, which has been grafted into this branch.  I'd prefer the existing login route take an `async` argument in the request and return a json blob instead of the redirect, leaving the rest of the handling intact.

**Comment by guerler** [comment]
Yeah, the arguments in #6512 are still valid. I am fine with reducing the toolshed code path separation within this PR.

File: client/galaxy/scripts/components/login/Login.vue
Comment (VJalili) [inline client/galaxy/scripts/components/login/Login.vue]: Not many users know what is an OpenID URL, or where to find it. Also, there is not different URLs for an identity provider. We would still be safe relying on the URLs we know; i.e., using the following drop-down where user can choose an OpenID. Therefore, I suggest removing this textbox and show the dropdown only.

File: client/galaxy/scripts/components/login/Login.vue
Comment (VJalili) [inline client/galaxy/scripts/components/login/Login.vue]: Would it be useful to add a note here saying something like "this feature will be deprecated in 19.05 release"?

File: client/galaxy/scripts/components/login/Login.vue
Comment (VJalili) [inline client/galaxy/scripts/components/login/Login.vue]: 1- "unkown" -> "unknown". 
2- "for unknown reason." -> "for an unknown reason."

File: client/galaxy/scripts/components/login/Login.vue
Comment (VJalili) [inline client/galaxy/scripts/components/login/Login.vue]: 1- "unkown" -> "unknown".
2- "for unknown reason." -> "for an unknown reason."

File: client/galaxy/scripts/components/login/Password.vue
Comment (VJalili) [inline client/galaxy/scripts/components/login/Password.vue]: 1- "unkown" -> "unknown".
2- "for unknown reason." -> "for an unknown reason."

File: lib/galaxy/auth/__init__.py
Comment (VJalili) [inline lib/galaxy/auth/__init__.py:101]: If I read this correct, you return `void` when password change is allowed, and return a message when it is not. I think you decided not to use boolean, because you have two different reasons for "False". However, I think returning a tuple (`boolean, message`) is intuitive. Would it be a challenge to keep it as it is now?

File: lib/galaxy/auth/__init__.py
Comment (VJalili) [inline lib/galaxy/auth/__init__.py]: Is the `kwd` used here?

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py]: I think `kwd` is not used. If so, could you plz remove it? If you remove it, you may need to update line #267 as well.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py:566]: You're still inferring a boolean "status" by `status is None`.  Therefore, I think the combination of (boolean)`success` + (string)`message` was more intuitive than the new combination of (string)`status` + (string)`message`.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py]: Would not it be better to default the payload to `{}`?

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py:603]: I think a documentation of arguments, payload in particular, would be quite useful here.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py:622]: I think a documentation of arguments, payload in particular, would be quite useful here. 

Also, I think you're not using `**kwd`; if so, would you mind please remove it?

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py]: Is this `message` used anywhere?!

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py]: `params` is not used, if it is not needed, I would recommend removing it.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [inline lib/galaxy/webapps/galaxy/controllers/user.py:635]: Would it be better to default `redirect` to `None`? 
and at line 683 check for `None` instead of `"None"`?

File: templates/user/openid_manage.mako
Comment (VJalili) [inline templates/user/openid_manage.mako:9]: Do you think putting a text such as "this feature will be deprecated in 19.05 release" would be useful?

File: templates/user/openid_manage.mako
Comment (VJalili) [inline templates/user/openid_manage.mako]: Similar to my other comment, I think not many users know what is an OpenID URL, or where to find it. Also, there is not different URLs for an identity provider (it is usually only one). We would still be safe relying on the URLs we know; i.e., using the following drop-down where user can choose an OpenID. Therefore, I suggest removing this textbox and show the dropdown only.

**Review by VJalili — CHANGES_REQUESTED** [review]

**Comment by VJalili** [comment]
@guerler, I very much liked that you moved all the logics from the API to their manager. Thanks.

This looks good to me. I added few minor comments. 

One point about how it looks on screen: (I don't know if it is on my end only, or it is the intended behavior) I see text boxes span as wide as window size (see the following screenshot). I think, it would be more elegant to centralize them horizontally and vertically, and limit their width to a much smaller width. 

screen shot 2018-07-30 at 12 05 59 pm

File: client/galaxy/scripts/components/login/Login.vue
Comment (guerler) [reply client/galaxy/scripts/components/login/Login.vue]: Done.

File: client/galaxy/scripts/components/login/Login.vue
Comment (guerler) [reply client/galaxy/scripts/components/login/Login.vue]: Done.

File: client/galaxy/scripts/components/login/Login.vue
Comment (guerler) [reply client/galaxy/scripts/components/login/Login.vue]: Done.

File: client/galaxy/scripts/components/login/Login.vue
Comment (guerler) [reply client/galaxy/scripts/components/login/Login.vue]: Done.

File: client/galaxy/scripts/components/login/Password.vue
Comment (guerler) [reply client/galaxy/scripts/components/login/Password.vue]: Done.

File: lib/galaxy/auth/__init__.py
Comment (guerler) [reply lib/galaxy/auth/__init__.py:101]: This function is only called once within the same module. It's trivial to change it but I am not sure how useful it would be.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (guerler) [reply lib/galaxy/webapps/galaxy/controllers/user.py]: Done.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (guerler) [reply lib/galaxy/webapps/galaxy/controllers/user.py]: Done.

File: templates/user/openid_manage.mako
Comment (guerler) [reply templates/user/openid_manage.mako:9]: Done.

File: templates/user/openid_manage.mako
Comment (guerler) [reply templates/user/openid_manage.mako]: Done.

File: lib/galaxy/auth/__init__.py
Comment (guerler) [reply lib/galaxy/auth/__init__.py]: Done.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (guerler) [reply lib/galaxy/webapps/galaxy/controllers/user.py]: Done.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (guerler) [reply lib/galaxy/webapps/galaxy/controllers/user.py]: Done.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (guerler) [reply lib/galaxy/webapps/galaxy/controllers/user.py:566]: This is temporary, in the follow-up of this PR the Registration mako will be replaced and this status parameter will be dropped.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (guerler) [reply lib/galaxy/webapps/galaxy/controllers/user.py]: Done.

File: lib/galaxy/auth/__init__.py
Comment (VJalili) [reply lib/galaxy/auth/__init__.py:101]: I think it would make it more intuitive; however, it I would leave it up to you to decide.

File: lib/galaxy/webapps/galaxy/controllers/user.py
Comment (VJalili) [reply lib/galaxy/webapps/galaxy/controllers/user.py:566]: Sounds good. Thanks.

**Comment by guerler** [comment]
ping @VJalili

File: lib/galaxy/managers/users.py
Comment (VJalili) [inline lib/galaxy/managers/users.py:265]: IMO the second `not` is not required.

File: lib/galaxy/managers/users.py
Comment (VJalili) [inline lib/galaxy/managers/users.py:305]: Since you're updating an existing record, you may not need to `add` it to the database, though it needs to be tested. The following `flush` might be enough to commit your updates.

**Comment by VJalili** [comment]
- As I mentioned in the previous comment, the login page spans whole browser width; maybe it would be better to limit its width and align them horizontally to the center of screen. 
- In the register page, some UI elements have a maximum width, but some span throughout whole window (see the following screenshot); I think all elements should have a maximum width and centered horizontally. 

screen shot 2018-08-24 at 3 35 20 pm

- Would not it be useful to show new password strength when changing a current password?

**Comment by guerler** [comment]
@VJalili thanks for the review and your comments. I revised the page to limit the width for input elements through columns. The login page now looks as shown belown. The latter two comments are related to the registration mako replacement PR #6621 and I will revisit them there once this PR is merged. I think we should merge this PR after the upcoming freeze, early in the release cycle.

loginpage

**Comment by VJalili** [comment]
@guerler Thanks for updating this. It looks much nicer now. 

Also, I agree that this PR and #6621 need to be merged as soon as possible (ideally for the next release), since we do have some parallel work that depend on the functionality you implemented in these two PRs.

**Review by VJalili — APPROVED** [review]

**Comment by dannon** [comment]
@VJalili I think @guerler was suggesting the opposite, that we merge this *after* the freeze and not before.

**Comment by VJalili** [comment]
Thats correct. Indeed I was wondering if that would be possible to merge this soon so that we can have it in the next release.

**Comment by VJalili** [comment]
@guerler is this ready to be reviewed and merged?

**Pull request closed without merging**
